pkg/rbd/rbd_attach.go

```diff
@@ -0,0 +1,303 @@
+		if _, err := os.Lstat(devicePath); err != nil {
+			glog.Warningf("Stat device %s for imgpath %s failed %v", devicePath, imgPath, err)
+			continue
+		}
+		return devicePath, true
+	}
+	return "", false
+}
+
+// Stat a path, if it doesn't exist, retry maxRetries times.
+func waitForPath(pool, image string, maxRetries int, useNbdDriver bool) (string, bool) {
+	for i := 0; i < maxRetries; i++ {
+		if i != 0 {
+			time.Sleep(time.Second)
+		}
+		if useNbdDriver {
+			if devicePath, found := getNbdDevFromImageAndPool(pool, image); found {
+				return devicePath, true
+			}
+		} else {
+			if devicePath, found := getRbdDevFromImageAndPool(pool, image); found {
+				return devicePath, true
+			}
+		}
+	}
+	return "", false
+}
+
+// Check if rbd-nbd tools are installed.
+func checkRbdNbdTools() bool {
+	_, err := execCommand("modprobe", []string{"nbd"})
+	if err != nil {
+		glog.V(3).Infof("rbd-nbd: nbd modprobe failed with error %v", err)
+		return false
+	}
+	if _, err := execCommand("rbd-nbd", []string{"--version"}); err != nil {
+		glog.V(3).Infof("rbd-nbd: running rbd-nbd -h failed with error %v", err)
+		return false
+	}
+	glog.V(3).Infof("rbd-nbd tools were found.")
+	return true
+}
+
+func attachRBDImage(volOptions *rbdVolume, userId string, credentials map[string]string) (string, error) {
+	var err error
+	var output []byte
+
+	image := volOptions.VolName
+	imagePath := fmt.Sprintf("%s/%s", volOptions.Pool, image)
+
+	cmdName := "rbd"
+	moduleName := "rbd"
+	if useNBD {
+		cmdName = "rbd-nbd"
+		moduleName = "nbd"
+	}
+	devicePath, found := waitForPath(volOptions.Pool, image, 1, useNBD)
+	if !found {
+		attachdetachMutex.LockKey(string(volOptions.Pool + image))
+		defer attachdetachMutex.UnlockKey(string(imagePath))
+
+		_, err = execCommand("modprobe", []string{moduleName})
+		if err != nil {
+			glog.Warningf("rbd: failed to load rbd kernel module:%v", err)
+		}
+
+		backoff := wait.Backoff{
+			Duration: rbdImageWatcherInitDelay,
+			Factor:   rbdImageWatcherFactor,
+			Steps:    rbdImageWatcherSteps,
+		}
+		err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+			used, rbdOutput, err := rbdStatus(volOptions, userId, credentials)
+			if err != nil {
+				return false, fmt.Errorf("fail to check rbd image status with: (%v), rbd output: (%s)", err, rbdOutput)
+			}
+			return !used, nil
+		})
+		// return error if rbd image has not become available for the specified timeout
+		if err == wait.ErrWaitTimeout {
+			return "", fmt.Errorf("rbd image %s is still being used", imagePath)
+		}
+		// return error if any other errors were encountered during wating for the image to becme avialble
+		if err != nil {
+			return "", err
+		}
+
+		glog.V(3).Infof("rbd: map mon %s", volOptions.Monitors)
+		key, err := getRBDKey(userId, credentials)
+		if err != nil {
+			return "", err
+		}
+		output, err = execCommand(cmdName, []string{
+			"map", imagePath, "--id", userId, "-m", volOptions.Monitors, "--key=" + key})
```

Comment:
Nit: I know this is an existing issue, but it doesn't seem safe to be passing the key in clear text. You can use the ```--keyfile``` optional to pass a path to a protected file which contains the key instead.

Reply:
yes, this is a long standing issue, especially in this case, rbdplugin has to pass /proc/pid/cmdline and get a chance to view the secret.

a potential fix is to let `rbd-nbd` print out the device path, so rbdplugin doesn't have to probe each nbd device through /proc to figure out which device maps to the image. I'll open a ceph issue for this feature.